File: snippets/go/nostr_wallet_connect.go
package example

import (
	"context"
	"fmt"
	"log"

	breez_sdk_liquid "github.com/breez/breez-sdk-liquid-go/breez_sdk_liquid"
)

// ANCHOR: event-listener
type MyNwcEventListener struct{}

func (l *MyNwcEventListener) OnEvent(ctx context.Context, event breez_sdk_liquid.NwcEvent) error {
	switch details := event.Details.(type) {
	case breez_sdk_liquid.NwcEventDetailsConnected:
		fmt.Println("NWC connected")
	case breez_sdk_liquid.NwcEventDetailsDisconnected:
		fmt.Println("NWC disconnected")
	case breez_sdk_liquid.NwcEventDetailsPayInvoice:
		if details.Success {
			fmt.Println("Payment successful")
		} else {
			fmt.Println("Payment failed")
		}
	case breez_sdk_liquid.NwcEventDetailsListTransactions:
		fmt.Println("Transactions requested")
	case breez_sdk_liquid.NwcEventDetailsGetBalance:
		fmt.Println("Balance requested")
	}
	return nil
}

// ANCHOR_END: event-listener

func nostrWalletConnect() error {
	// ANCHOR: nwc-config
	nwcConfig := breez_sdk_liquid.NwcConfig{
		RelayUrls:    []string{"<your-relay-url-1>"}, // Optional: Custom relay URLs (uses default if nil)
		SecretKeyHex: "your-nostr-secret-key-hex",    // Optional: Custom Nostr secret key
	}

	nwcService := breez_sdk_liquid.NewBindingNwcService(nwcConfig)

	// Add the plugin to your SDK
	plugins := []breez_sdk_liquid.Plugin{nwcService}
	// ANCHOR_END: nwc-config
	_ = plugins

	// ANCHOR: add-connection
	connectionName := "my-app-connection"
	connectionString, err := nwcService.AddConnectionString(connectionName)
	if err != nil {
		return err
	}
	// ANCHOR_END: add-connection

	// ANCHOR: list-connections
	connections, err := nwcService.ListConnectionStrings()
	if err != nil {
		return err
	}
	// ANCHOR_END: list-connections
	_ = connections

	// ANCHOR: remove-connection
	err = nwcService.RemoveConnectionString(connectionName)
	if err != nil {
		return err
	}
	// ANCHOR_END: remove-connection

	// ANCHOR: event-management
	// Add event listener
	listener := &MyNwcEventListener{}
	listenerId, err := nwcService.AddEventListener(listener)
	if err != nil {
		return err
	}

	// Remove event listener when no longer needed
	err = nwcService.RemoveEventListener(listenerId)
	if err != nil {
		return err
	}
	// ANCHOR_END: event-management

	// ANCHOR: error-handling
	connectionString, err = nwcService.AddConnectionString("test")
	if err != nil {
		switch e := err.(type) {
		case breez_sdk_liquid.NwcErrorGeneric:
			log.Printf("Generic error: %s", e.Message)
		case breez_sdk_liquid.NwcErrorPersist:
			log.Printf("Persistence error: %s", e.Message)
		default:
			log.Printf("Unknown error: %v", err)
		}
		return err
	}
	fmt.Printf("Connection created: %s", connectionString)
	// ANCHOR_END: error-handling

	return nil
}

File: snippets/rust/src/nostr_wallet_connect.rs
use breez_sdk_liquid::plugin::Plugin;
use breez_sdk_liquid_nwc::{
    NwcConfig,
    SdkNwcService,
    error::NwcError,
    event::{NwcEvent, NwcEventDetails, NwcEventListener},
    NwcService,
};
use std::sync::Arc;
use anyhow::Result;
use async_trait::async_trait;

async fn nostr_wallet_connect() -> Result<()> {
    // ANCHOR: nwc-config
    let nwc_config = NwcConfig {
        relay_urls: Some(vec!["<your-relay-url-1>".to_string()]),          // Optional: Custom relay URLs (uses default if None)
        secret_key_hex: Some("your-nostr-secret-key-hex".to_string()),     // Optional: Custom Nostr secret key
    };
    
    let nwc_service = Arc::new(SdkNwcService::new(nwc_config));
    
    // Add the plugin to your SDK
    let plugins: Vec<Arc<dyn Plugin>> = vec![nwc_service.clone()];
    // ANCHOR_END: nwc-config

    // ANCHOR: add-connection
    let connection_name  = "my-app-connection";
    let connection_string = nwc_service.add_connection_string(connection_name.to_string()).await?;
    // ANCHOR_END: add-connection

    // ANCHOR: list-connections
    let connections = nwc_service.list_connection_strings().await?;
    // ANCHOR_END: list-connections

    // ANCHOR: remove-connection
    nwc_service.remove_connection_string(connection_name.to_string()).await?;
    // ANCHOR_END: remove-connection

    // ANCHOR: event-listener
    struct MyNwcEventListener;

    #[async_trait]
    impl NwcEventListener for MyNwcEventListener {
        async fn on_event(&self, event: NwcEvent) {
            match event.details {
                NwcEventDetails::Connected => println!("NWC connected"),
                NwcEventDetails::Disconnected => println!("NWC disconnected"),
                NwcEventDetails::PayInvoice { success, .. } => {
                    println!("Payment {}", if success { "successful" } else { "failed" });
                }
                NwcEventDetails::ListTransactions => println!("Transactions requested"),
                NwcEventDetails::GetBalance => println!("Balance requested"),
            }
        }
    }
    // ANCHOR_END: event-listener

    // ANCHOR: event-management
    // Add event listener
    let listener = Box::new(MyNwcEventListener);
    let listener_id = nwc_service.add_event_listener(listener).await;

    // Remove event listener when no longer needed
    nwc_service.remove_event_listener(&listener_id).await;
    // ANCHOR_END: event-management

    // ANCHOR: error-handling
    match nwc_service.add_connection_string("test".to_string()).await {
        Ok(connection_string) => println!("Connection created: {}", connection_string),
        Err(NwcError::Generic { err }) => eprintln!("Generic error: {}", err),
        Err(NwcError::Persist { err }) => eprintln!("Persistence error: {}", err),
    }
    // ANCHOR_END: error-handling
    
    Ok(())
}